### DISCUSSION TITLE
Use for automatic differentiation?

### DISCUSSION THREAD

**Discussion by rloekvh** [mutationpp/Mutationpp#182] [Ideas]
Hi all, 

I was wondering if there would be any interest or benefit to applying automatic differentiation (AD) to Mutation++? 

We’re attempting to use Mutation inside our DG CFD code for simulation of nonequilibrium flow. One goal of ours is to use automatic differentiation with [Enzyme](https://enzyme.mit.edu) to propagate gradients through our code. We’re interested in gradient-based sensitivity analysis, but differentiable code has many other uses; the Enzyme [paper](https://arxiv.org/pdf/2010.01709.pdf) has an example of using Enzyme within TensorFlow. I'm not sure how robust this support is, but the plugin is available [here](https://github.com/wsmoses/Enzyme-TF).

At some point, the gradients will propagate down to Mutation++ calls. I know that there are analytical derivatives in JacobianManager and elsewhere in the code, but are there gaps in the derivative information, for example for transport algorithms, that could be filled by AD? I know this is a bit vague, but I'm interested to hear people's thoughts whether a "differentiable" Mutation++ would be a helpful tool or whether it's unnecessary or potentially problematic.

**Comment by ggange** [comment]
Hello @rloekvh, sorry for this very late reply! To my knowledge at the moment there are no efforts in including AD in Mutation++ but I know other projects have used the library coupling it to AD solvers, with very nice results; I don't know how helpful it will be for you but here is a reference (https://www.mdpi.com/2226-4310/8/11/322/htm) ;)
